feat: Implement EC2 Key Pairs (#1403)

jhudson10x opened the pull request:
<!-- 🎉 Thank you for making CloudQuery awesome by submitting a PR 🎉 -->

#### Summary

Add new resource for EC2 Key Pairs

https://github.com/cloudquery/cq-provider-aws/issues/1403

---

Use the following steps to ensure your PR is ready to be reviewed

- [X] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [X] Run `go fmt` to format your code 🖊
- [X] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [X] Update or add tests. Learn more about testing [here](https://docs.cloudquery.io/docs/developers/sdk/testing) 🧪
- [X] Update the docs by running `go run ./docs/docs.go` and committing the changes 📃
- [ ] Ensure the status checks below are successful ✅

jhudson10x committed:
feat: Implement EC2 Key Pairs (#1403)

jhudson10x committed:
Register key_pairs in the provider config

jhudson10x committed:
Fix table configuration

jhudson10x committed:
Update docs

candiduslynx commented on resources/services/ec2/key_pairs.go:
```diff
@@ -0,0 +1,78 @@
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+)
+
+func Ec2KeyPairs() *schema.Table {
```
Kindly add `cq-gen` config for this resource.
Additionally, the generate instruction comment should be present here, and no diff should occur once you run this generate command on this file.

See [AWS SES Template resource](https://github.com/cloudquery/cq-provider-aws/blob/main/resources/services/ses/templates.go#L13) for inspiraition.

jhudson10x replied:
I will add an entry to the `resources/services/ec2/gen.hcl` file for this resource. I notice that there are other resources with missing entries in this file. Can you provide me more info about the purpose of this file?

candiduslynx replied:
We tend to use `cq-gen` to generate resources (may be installed via `make install-tools`).
Additionally, it provides an easy way to specify the resource structure.

Also, we are going to check for drift changes for the generated resources (so you should be able to run `go generate` on this file and see no changes whatsoever).

candiduslynx commented on resources/services/ec2/key_pairs.go:
```diff
@@ -0,0 +1,78 @@
+		Description:   "Describes an EC2 Key Pair.",
+		Resolver:      fetchEc2KeyPairs,
+		Multiplex:     client.ServiceAccountRegionMultiplexer("ec2"),
+		IgnoreError:   client.IgnoreWithInvalidAction,
```
Why did you choose this specific ignore function?
Will `client.client.IgnoreAccessDeniedServiceDisabled` work?

jhudson10x replied:
I don't have a good reason for having chosen that specific ignore function. I will change it to client.IgnoreAccessDeniedServiceDisabled. If there is an even better ignore handler to use for this situation tell me and I will change it again.

candiduslynx commented on resources/services/ec2/key_pairs.go:
```diff
@@ -0,0 +1,78 @@
+		Multiplex:     client.ServiceAccountRegionMultiplexer("ec2"),
+		IgnoreError:   client.IgnoreWithInvalidAction,
+		DeleteFilter:  client.DeleteAccountRegionFilter,
+		Options:       schema.TableCreationOptions{PrimaryKeys: []string{"key_name"}},
```
Use `key_pair_id` instead

jhudson10x committed:
Incorporate PR feedback

- Add entry to gen.hcl for KeyPairInfo
- Add generate instruction comment to func Ec2KeyPairs()
- Change error handler to client.IgnoreAccessDeniedServiceDisabled
- Use key_pair_id for table primary key

candiduslynx commented on resources/services/ec2/gen.hcl:
```diff
@@ -249,4 +249,44 @@ resource "aws" "ec2" "instance_types" {
+    primary_keys = ["key_pair_id"]
+  }
+
+  userDefinedColumn "account_id" {
```
does Key Pair have ARN?

candiduslynx replied:
Then you should add a resolver for it, too.

Basically, you should add (and specify this column to e primary key, see lots of examples in the reources):
```hcl
userDefinedColumn "arn" {
    type              = "string"
    description       = "The Amazon Resource Name (ARN) of the Key Pair"
    generate_resolver = true
}
```
Then you'll need to regenerate the implementation and implement the newly generated resolver.
You can find examples of resolvers utilizing helper functions like [this one](https://github.com/cloudquery/cq-provider-aws/blob/main/client/helpers.go#L290) throughout the codebase.

jhudson10x replied:
The func signature that cq-gen has a return type of `error`. Should I change the sig to return `schema.ColumnResolver` like the example you linked, or should I keep the generated sig and do like `resources/services/iot/iot_topic_rules.go#L1337` and write into the resource?

candiduslynx replied:
You should do something like https://github.com/cloudquery/cq-provider-aws/blob/main/resources/services/ec2/images.go#L42-L44

I suppose you can just fill in the proper func body for the generated func.

erezrokah commented:
Hi @jhudson10x and @candiduslynx, I moved this PR to our monorepo (including original commits) via https://github.com/cloudquery/cloudquery/pull/1325. Since https://github.com/cloudquery/cloudquery/pull/1325 was approved and merged I suggest to do any follow up changes in another PR.
Thanks for the contribution @jhudson10x!

erezrokah commented:
Closing in favor https://github.com/cloudquery/cloudquery/pull/1325
